An unhandled exception was thrown by the Error Prone static analysis plugin.

### What version of Error Prone are you using?
2.2.0

### Does this issue reproduce with the latest release?
Yes

### What did you do?
[InputStream.java.zip](https://github.com/google/error-prone/files/1764293/InputStream.java.zip)
./gradlew assembleDebug

### What did you expect to see?
No exception thrown by the Error Prone static analysis plugin

### What did you see instead?
```
InputStream.java:6: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
public class InputStream extends java.io.InputStream {
       ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.2.0
     Stack Trace:
     java.lang.NullPointerException
        at com.google.errorprone.bugpatterns.InputStreamSlowMultibyteRead.maybeMatchReadByte(InputStreamSlowMultibyteRead.java:114)
        at com.google.errorprone.bugpatterns.InputStreamSlowMultibyteRead.matchClass(InputStreamSlowMultibyteRead.java:108)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:581)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:145)
        at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:602)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:145)
        at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:64)
        at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
        at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1353)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:952)
        at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
        at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
        at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:137)
        at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:108)
        at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:118)
        at com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:65)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at net.ltgt.gradle.errorprone.ErrorProneCompiler.execute(ErrorProneCompiler.java:66)
        at net.ltgt.gradle.errorprone.ErrorProneCompiler.execute(ErrorProneCompiler.java:23)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:99)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:52)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:37)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilationFinalizer.execute(IncrementalCompilationFinalizer.java:39)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilationFinalizer.execute(IncrementalCompilationFinalizer.java:24)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:198)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:129)
        at com.android.build.gradle.tasks.factory.AndroidJavaCompile.compile(AndroidJavaCompile.java:95)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:173)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)
1 error
7 warnings
```

--- a/core/src/main/java/com/google/errorprone/bugpatterns/CatchAndPrintStackTrace.java
+++ b/core/src/main/java/com/google/errorprone/bugpatterns/CatchAndPrintStackTrace.java
@@ -43,9 +43,7 @@ public class CatchAndPrintStackTrace extends BugChecker implements CatchTreeMatc
 
   private static final Matcher<StatementTree> MATCHER =
       expressionStatement(
-          instanceMethod()
-              .onDescendantOf("java.lang.Throwable")
-              .withSignature("printStackTrace()"));
+          instanceMethod().onDescendantOf("java.lang.Throwable").named("printStackTrace"));
 
   @Override
   public Description matchCatch(CatchTree tree, VisitorState state) {

--- a/core/src/main/java/com/google/errorprone/bugpatterns/InputStreamSlowMultibyteRead.java
+++ b/core/src/main/java/com/google/errorprone/bugpatterns/InputStreamSlowMultibyteRead.java
@@ -111,12 +111,15 @@ public class InputStreamSlowMultibyteRead extends BugChecker implements ClassTre
   private Description maybeMatchReadByte(MethodTree readByteMethod, VisitorState state) {
     // Methods that return a constant expression are likely to be 'dummy streams', for which
     // the multibyte read is OK.
-    List<? extends StatementTree> statements = readByteMethod.getBody().getStatements();
-    if (statements.size() == 1) {
-      Tree tree = statements.get(0);
-      if (tree.getKind() == Kind.RETURN
-          && ASTHelpers.constValue(((ReturnTree) tree).getExpression()) != null) {
-        return Description.NO_MATCH;
+
+    if (readByteMethod.getBody() != null) { // Null-check for native/abstract overrides of read()
+      List<? extends StatementTree> statements = readByteMethod.getBody().getStatements();
+      if (statements.size() == 1) {
+        Tree tree = statements.get(0);
+        if (tree.getKind() == Kind.RETURN
+            && ASTHelpers.constValue(((ReturnTree) tree).getExpression()) != null) {
+          return Description.NO_MATCH;
+        }
       }
     }
 

--- a/core/src/test/java/com/google/errorprone/bugpatterns/CatchAndPrintStackTraceTest.java
+++ b/core/src/test/java/com/google/errorprone/bugpatterns/CatchAndPrintStackTraceTest.java
@@ -41,6 +41,12 @@ public class CatchAndPrintStackTraceTest {
             "      // BUG: Diagnostic contains:",
             "      t.printStackTrace();",
             "    }",
+            "    try {",
+            "      System.err.println();",
+            "    } catch (Throwable t) {",
+            "      // BUG: Diagnostic contains:",
+            "      t.printStackTrace(System.err);",
+            "    }",
             "  }",
             "}")
         .doTest();
@@ -61,11 +67,6 @@ public class CatchAndPrintStackTraceTest {
             "    } catch (Throwable t) {",
             "      t.printStackTrace();",
             "      t.printStackTrace();",
-            "    }",
-            "    try {",
-            "      System.err.println();",
-            "    } catch (Throwable t) {",
-            "      t.printStackTrace(System.err);",
             "    }",
             "  }",
             "}")

--- a/core/src/test/java/com/google/errorprone/bugpatterns/InputStreamSlowMultibyteReadTest.java
+++ b/core/src/test/java/com/google/errorprone/bugpatterns/InputStreamSlowMultibyteReadTest.java
@@ -47,7 +47,7 @@ public class InputStreamSlowMultibyteReadTest {
   }
 
   @Test
-  public void basic() throws Exception {
+  public void basic() {
     compilationHelper
         .addSourceLines(
             "TestClass.java",
@@ -55,6 +55,30 @@ public class InputStreamSlowMultibyteReadTest {
             "  byte[] buf = new byte[42];",
             "  // BUG: Diagnostic contains:",
             "  public int read() { return buf[0]; }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void abstractOverride() {
+    compilationHelper
+        .addSourceLines(
+            "TestClass.java",
+            "abstract class TestClass extends java.io.InputStream {",
+            "  // BUG: Diagnostic contains:",
+            "  public abstract int read();",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nativeOverride() {
+    compilationHelper
+        .addSourceLines(
+            "TestClass.java",
+            "abstract class TestClass extends java.io.InputStream {",
+            "  // BUG: Diagnostic contains:",
+            "  public native int read();",
             "}")
         .doTest();
   }

--- a/core/src/test/java/com/google/errorprone/bugpatterns/StreamResourceLeakTest.java
+++ b/core/src/test/java/com/google/errorprone/bugpatterns/StreamResourceLeakTest.java
@@ -204,6 +204,29 @@ public class StreamResourceLeakTest {
   }
 
   @Test
+  public void returnFromMustBeClosedMethodWithChaining() {
+    testHelper
+        .addSourceLines(
+            "Test.java",
+            "import com.google.errorprone.annotations.MustBeClosed;",
+            "import java.io.IOException;",
+            "import java.nio.file.Files;",
+            "import java.nio.file.Path;",
+            "import java.util.stream.Stream;",
+            "class Test {",
+            "  @MustBeClosed",
+            "  Stream<String> f(Path p) throws IOException {",
+            "    return Files.list(p).map(Path::toString); // OK due to @MustBeClosed",
+            "  }",
+            "  Stream<String> g(Path p) throws IOException {",
+            "    // BUG: Diagnostic contains: should be closed",
+            "    return Files.list(p).map(Path::toString);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
   public void moreRefactorings() throws IOException {
     BugCheckerRefactoringTestHelper.newInstance(new StreamResourceLeak(), getClass())
         .addInputLines(

--- a/docs/bugpattern/CatchAndPrintStackTrace.md
+++ b/docs/bugpattern/CatchAndPrintStackTrace.md
@@ -25,9 +25,3 @@ try {
   e.printStackTrace();
 }
 ```
-
-If you truly intend to print a stack trace to stderr, do so explicitly:
-
-```java
-e.printStackTrace(System.err);
-```

